layers: Add VK_EXT_host_query_reset validations

rg3igalia opened the pull request:
Add validations for VK_EXT_host_query_reset VUIDs and tests for them.

mark-lunarg commented on layers/core_validation.cpp:
```diff
@@ -12888,6 +12898,52 @@ bool CoreChecks::PreCallValidateGetBufferDeviceAddressEXT(VkDevice device, const
+
+bool CoreChecks::PreCallValidateResetQueryPoolEXT(VkDevice device, VkQueryPool queryPool, uint32_t firstQuery,
+                                                  uint32_t queryCount) {
+    bool skip = false;
```
This all looks absolutely fabulous.  I would add a query_validation check here, in case somebody wants to skip query validation -- something like    ` if (disabled.query_validation) return false;`

rg3igalia replied:
Thanks for the feedback! I added the check and force-pushed again.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -4910,6 +4915,11 @@ void CoreChecks::PostCallRecordCreateQueryPool(VkDevice device, const VkQueryPoo
+
+    for (uint32_t i = 0; i < pCreateInfo->queryCount; ++i) {
+        QueryObject query{*pQueryPool, i};
+        queryToStateMap[query] = false;
```
@tobine -- if we're going to initialize the queryToStateMap on creation, wouldn't it make sense to convert it to simple std::vector member of the `QUERY_POOL_NODE' -- it's not going to save a hash until/unless we "dehandlize" (i.e. store a shared_ptr<QUERY_POOL_NODE> inside of the QueryObject meta-handle, but once we do, it becomes a lookup.

(not for this PR, obviously)

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -12888,6 +12898,54 @@ bool CoreChecks::PreCallValidateGetBufferDeviceAddressEXT(VkDevice device, const
+                        firstQuery, totalCount, report_data->FormatHandle(queryPool).c_str());
+    }
+
+    if (firstQuery + queryCount > totalCount) {
```
style preference is for () around mathematical operations with results used in comparison.

rg3igalia replied:
Will fix.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -12888,6 +12898,54 @@ bool CoreChecks::PreCallValidateGetBufferDeviceAddressEXT(VkDevice device, const
+                        "VUID-vkResetQueryPoolEXT-None-02665", "Host query reset not enabled for device");
+    }
+
+    auto i = queryPoolMap.find(queryPool);
```
i is a terrible name, :(.  `query_pool_it` or just `query_pool` would be better.

rg3igalia replied:
Will fix.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -12888,6 +12898,54 @@ bool CoreChecks::PreCallValidateGetBufferDeviceAddressEXT(VkDevice device, const
+    return skip;
+}
+
+void CoreChecks::PostCallRecordResetQueryPoolEXT(VkDevice device, VkQueryPool queryPool, uint32_t firstQuery, uint32_t queryCount) {
```
@mark-lunarg  @tobine  -- do we want to reset the query state if the feature *isn't* supported?

mark-lunarg replied:
This is not the _feature_, this is the feature _validation_.  So, yes, all state gets all updated always.

jzulauf-lunarg replied:
I was thinking of `core_checks->enabled_features.host_query_reset_features` not `disabled.query_validation`

jzulauf-lunarg replied:
Additionally, we might want to check that queryPool has an entry in queryPoolMap, s.t. we aren't setting entries in the queryToStateMap, that nobody owns and no-one will clean up.

rg3igalia replied:
Will fix both things. I agree. If the feature is not enabled it may be better not to reset the query state at all.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -12888,6 +12898,54 @@ bool CoreChecks::PreCallValidateGetBufferDeviceAddressEXT(VkDevice device, const
+
+void CoreChecks::PostCallRecordResetQueryPoolEXT(VkDevice device, VkQueryPool queryPool, uint32_t firstQuery, uint32_t queryCount) {
+    for (uint32_t i = 0; i < queryCount; ++i) {
+        QueryObject query{queryPool, firstQuery + i};
```
several choices to not recreate the QueryObject each time (including using QueryObject::index (or a reference to it) as the loop index

--or-- going the other way...

`queryToStateMap[QueryObject(queryPool, firstQuery + 1)] = false;`

jzulauf-lunarg replied:
In line with the comment above about queryToStateMap entry lifetimes, we don't want to be creating entries here so either we make sure the query pool exists above, or we use `find` instead of `operator []` here.

rg3igalia replied:
> In line with the comment above about queryToStateMap entry lifetimes, we don't want to be creating entries here so either we make sure the query pool exists above, or we use `find` instead of `operator []` here.

Will fix.

jzulauf-lunarg commented on layers/core_validation.cpp:
```diff
@@ -12888,6 +12898,54 @@ bool CoreChecks::PreCallValidateGetBufferDeviceAddressEXT(VkDevice device, const
+
+    bool skip = false;
+
+    if (!GetEnabledFeatures()->host_query_reset_features.hostQueryReset) {
```
Helper function deleted at https://github.com/KhronosGroup/Vulkan-ValidationLayers/commit/d7b03cc3f94e443dc1fea38fc97c6b82be77540c rebase/use direct reference to build against master.

rg3igalia replied:
Will fix.

rg3igalia commented:
I think I've addressed all feedback reported above, but let me know if there's any detail missing. Regarding "memory leaks" and other query state tracking issues:

In this PR, I've only tried to get validations working for VK_EXT_host_query_reset, but I did find some of the issues mentioned above and then some.

* There are per-command-buffer, per-queue and a global queryToStateMap maps. The global one gets new entries when work is retired from queues (see CoreChecks::RetireWorkOnQueue) and I believe entries in it are never deleted, but I think the map is needed because queries can be accessed accross queues and command buffers. Query state tracking is complex.

* The current boolean state of queries is not enough to track correct query usage. For example, when queries are created they are in an undefined state and must be reset before use, as per the spec.

* CoreChecks::PostCallRecordGetQueryPoolResults does a complex dance to set a "needsSignaled" boolean value that I see is set in 3 places but never used for anything else. Am I missing something?

* There are subtle query state aspects the validation layers currently are not checking or cannot actually be checked properly. For example, after ending a query results are not actually available until the query is "finished", which must happen some finite time after the query is ended, and you must use the WAIT_BIT flag to guarantee you'll get results.

I have some changes that address some (but not all!) of the problems above and try to improve query state tracking. I can submit a PR with them once this one is merged.

rg3igalia commented:
@jzulauf-lunarg Friendly ping. I rebased and I think I addressed all your comments above. Let me know if there's anything else missing.

jzulauf-lunarg commented:
@rg3igalia -- looks good. layer validation tests refactor requires resolution. Just add your new tests to the appropriate (vklayertests.cpp or vkpositivelayertests.cpp) file.
